pinot-core/src/main/java/org/apache/pinot/core/operator/query/LinearSelectionOrderByOperator.java

```diff
@@ -0,0 +1,388 @@
+   */
+  protected static class TotallySortedListBuilder implements ListBuilder {
+    private final ArrayList<Object[]> _list;
+
+    public TotallySortedListBuilder(int expectedSize) {
+      _list = new ArrayList<>(expectedSize);
+    }
+
+    @Override
+    public boolean add(Object[] row) {
+      _list.add(row);
+      return true;
+    }
+
+    @Override
+    public List<Object[]> build() {
+      return _list;
+    }
+
+    @Override
+    public int sortedSize() {
+      return _list.size();
+    }
+  }
+
+  /**
+   * This is the more versatile {@link ListBuilder} that requires two comparators: The first defines the order on which
+   * {@link #add(Object[])} is called and defines the partitions. The second one is used to sort rows in the same
+   * paritition. This class can be used to implement order-by queries that include one or more not sorted expressions.
+   * In cases where all expressions are sorted, {@link TotallySortedListBuilder} should be used because its performance
+   * is better.
+   *
+   * In this implementation the {@link #add(Object[])} method has a log(P) cost, where P is the number of elements in
+   * each partition. In normal cases we can consider P a constant. In some degenerated cases like having an ordering by
+   * a sorted column whose values are all equal, P may tend to the number of elements in the segment, which would make
+   * the cost of this method log(N).
+   */
+  private static class PartiallySortedListBuilder implements ListBuilder {
+    private final ArrayList<Object[]> _rows;
+    /**
+     * The comparator that defines the partitions and the one that impose in which order add has to be called.
+     */
+    private final Comparator<Object[]> _partitionComparator;
+    /**
+     * The comparator that sorts different rows on each partition.
+     */
+    private final Comparator<Object[]> _unsortedComparator;
+    @Nullable
+    private Object[] _lastPartitionRow;
+    /**
+     * The index where the current (and still not sorted) partition starts.
+     */
+    private int _lastSorted;
+
+    public PartiallySortedListBuilder(int expectedSize, Comparator<Object[]> partitionComparator,
+        Comparator<Object[]> unsortedComparator) {
+      _rows = new ArrayList<>(expectedSize);
+      _partitionComparator = partitionComparator;
+      _unsortedComparator = unsortedComparator;
+    }
+
+    @Override
+    public boolean add(Object[] row) {
+      boolean result;
+      if (_lastPartitionRow == null) {
+        _lastSorted = 0;
+        _lastPartitionRow = row;
+        result = false;
+      } else {
+        int cmp = _partitionComparator.compare(row, _lastPartitionRow);
+        if (cmp < 0) {
+          throw new IllegalArgumentException("Row " + Arrays.toString(row) + " is lower than previously added row"
+              + Arrays.toString(_lastPartitionRow));
+        }
+        if (cmp > 0) {
```

Comment:
In some corner case, we might gather too many rows with sorted columns the same value, and cause memory issue. We might want to use a priority queue and only keep the `LIMIT` rows

Reply:
It would make the code a little bit slower, but what you say makes sense. I'm gonna change it

Reply:
I've created a new version that in case more than <maxElements> are found, the last partition is stored in a PriorityQueue